tests/unittests/Foundation/NSNotificationQueueTests.mm

```diff
@@ -0,0 +1,206 @@
+    NSNotificationCenter* notificationCenter = [[NSNotificationCenter new] autorelease];
+    NSNotificationQueueTestHelper* helper = [[NSNotificationQueueTestHelper new] autorelease];
+
+    __block StrongId<NSNotificationQueue> queue;
+    [helper scheduleAndAwaitBlock:^{
+        queue.attach([[NSNotificationQueue alloc] initWithNotificationCenter:notificationCenter]);
+    }];
+
+    __block NSCondition* condition = [[NSCondition new] autorelease];
+    __block int counter = 0;
+    [notificationCenter addObserverForName:s_TestNotificationName
+                                    object:nil
+                                     queue:nil
+                                usingBlock:^(NSNotification* note) {
+                                    ASSERT_OBJCEQ(s_TestNotificationName, note.name);
+                                    [condition lock];
+                                    ++counter;
+                                    [condition broadcast];
+                                    [condition unlock];
+                                }];
+
+    [condition lock];
+
+    [helper scheduleAndAwaitBlock:^{
+        [queue enqueueNotification:[NSNotification notificationWithName:s_TestNotificationName object:nil] postingStyle:NSPostASAP];
+    }];
+
+    [condition waitUntilDate:[NSDate dateWithTimeIntervalSinceNow:20]];
+    [condition unlock];
+
+    ASSERT_EQ(1, counter);
+}
+
+TEST(NSNotificationQueue, PostWhenIdle) {
+    static NSString* s_TestNotificationName = @(GetTestFullName().c_str());
+    NSNotificationCenter* notificationCenter = [[NSNotificationCenter new] autorelease];
+    NSNotificationQueueTestHelper* helper = [[NSNotificationQueueTestHelper new] autorelease];
+
+    __block StrongId<NSNotificationQueue> queue;
+    [helper scheduleAndAwaitBlock:^{
+        queue.attach([[NSNotificationQueue alloc] initWithNotificationCenter:notificationCenter]);
+    }];
+
+    __block NSCondition* condition = [[NSCondition new] autorelease];
+    __block int counter = 0;
+    [notificationCenter addObserverForName:s_TestNotificationName
+                                    object:nil
+                                     queue:nil
+                                usingBlock:^(NSNotification* note) {
+                                    ASSERT_OBJCEQ(s_TestNotificationName, note.name);
+                                    [condition lock];
+                                    ++counter;
+                                    [condition broadcast];
+                                    [condition unlock];
+                                }];
+
+    [condition lock];
+
+    [helper scheduleAndAwaitBlock:^{
+        [queue enqueueNotification:[NSNotification notificationWithName:s_TestNotificationName object:nil] postingStyle:NSPostWhenIdle];
+    }];
+
+    [condition waitUntilDate:[NSDate dateWithTimeIntervalSinceNow:20]];
+    [condition unlock];
+
+    ASSERT_EQ(1, counter);
+}
```

Comment:
Okay, but we'll only be testing that all three have fired; ordering is not guaranteed.